File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/1010-pairs-of-songs-divisible-by-60-approach2.go
package main

import "fmt"

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func numPairsDivisibleBy60(time []int) int {
	remainders := [60]int{}
	result := 0
	for _, t := range time {
		complement := (60 - t%60) % 60
		result += remainders[complement]
		remainders[t%60]++
	}
	return result
}

func runTests() {
	assert(numPairsDivisibleBy60([]int{30, 20, 150, 100, 40}) == 3, "test 1")
	assert(numPairsDivisibleBy60([]int{60, 60, 60}) == 3, "test 2")
	assert(numPairsDivisibleBy60([]int{10, 50, 90, 30}) == 2, "test 3")
	assert(numPairsDivisibleBy60([]int{1}) == 0, "test 4")
	assert(numPairsDivisibleBy60([]int{60}) == 0, "test 5")
	fmt.Println("all tests pass")
}

func main() { runTests() }

File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/1010-pairs-of-songs-divisible-by-60-approach2.py
def num_pairs_divisible_by60(time: list[int]) -> int:
    remainders = [0] * 60
    result = 0
    for t in time:
        complement = (60 - t % 60) % 60
        result += remainders[complement]
        remainders[t % 60] += 1
    return result

assert num_pairs_divisible_by60([30,20,150,100,40]) == 3
assert num_pairs_divisible_by60([60,60,60]) == 3
assert num_pairs_divisible_by60([10,50,90,30]) == 2
assert num_pairs_divisible_by60([1]) == 0
assert num_pairs_divisible_by60([60]) == 0
print("all tests pass")
